ipatests/test_integration/test_hsm.py

```diff
@@ -0,0 +1,1422 @@
+        """
+        Test that the KRA subsystem certificates renew properly
+        """
+        check_version(self.master)
+        CA_TRACKING_REQS = {
+            'ocspSigningCert cert-pki-ca': 'caocspSigningCert',
+            'subsystemCert cert-pki-ca': 'casubsystemCert',
+            'auditSigningCert cert-pki-ca': 'caauditSigningCert'
+        }
+        CA_TRACKING_REQS.update(KRA_TRACKING_REQS)
+        self.master.put_file_contents('/tmp/token_passwd', self.token_password)
+        for nickname in CA_TRACKING_REQS:
+            cert = tasks.certutil_fetch_cert(
+                self.master,
+                paths.PKI_TOMCAT_ALIAS_DIR,
+                '/tmp/token_passwd',
+                nickname,
+                token_name=self.token_name,
+            )
+            starting_serial = int(cert.serial_number)
+            cmd_arg = [
+                'ipa-getcert', 'resubmit', '-v', '-w',
+                '-d', paths.PKI_TOMCAT_ALIAS_DIR,
+                '-n', nickname,
+            ]
+            result = self.master.run_command(cmd_arg)
+            request_id = re.findall(r'\d+', result.stdout_text)
+
+            status = tasks.wait_for_request(self.master, request_id[0], 120)
+            assert status == "MONITORING"
+
+            args = ['-L', '-h', self.token_name, '-f', '/tmp/token_passwd']
+            tasks.run_certutil(self.master, args, paths.PKI_TOMCAT_ALIAS_DIR)
+
+            cert = tasks.certutil_fetch_cert(
+                self.master,
+                paths.PKI_TOMCAT_ALIAS_DIR,
+                '/tmp/token_passwd',
+                nickname,
+                token_name=self.token_name,
+            )
+            assert starting_serial != int(cert.serial_number)
+
+
+class TestHSMCALessToExternalToSelfSignedCA(CALessBase):
+    """Test server caless to extarnal CA to self signed scenario"""
+
+    num_replicas = 1
+
+    @classmethod
+    def install(cls, mh):
+        check_version(cls.master)
+        super(TestHSMCALessToExternalToSelfSignedCA, cls).install(mh)
+        # Enable pkiuser to read softhsm tokens
+        cls.master.run_command(['usermod', 'pkiuser', '-a', '-G', 'ods'])
+
+        token_name, token_passwd = get_hsm_token(cls.master)
+        cls.token_password = token_passwd
+        cls.token_name = token_name
+
+    @classmethod
+    def uninstall(cls, mh):
+        check_version(cls.master)
+        super(TestHSMCALessToExternalToSelfSignedCA, cls).uninstall(mh)
+        delete_hsm_token([cls.master] + cls.replicas, cls.token_name)
+
+    def test_hsm_caless_server(self):
+        """Install CA-less master"""
+        check_version(self.master)
+        self.create_pkcs12('ca1/server')
+        self.prepare_cacert('ca1')
+
+        master = self.install_server(
+            extra_args=[
+                '--token-name', self.token_name,
+                '--token-library-path', hsm_lib_path,
+                '--token-password', self.token_password
+            ]
```

Comment:
The extra args are not needed if the installation is CA-less. I would also expect the installer to complain about conflicting arguments.